compiler/rustc_mir_transform/src/generator.rs

```diff
@@ -775,34 +824,44 @@ fn compute_layout<'tcx>(
         storage_liveness,
     } = liveness;
 
-    // Gather live local types and their indices.
-    let mut locals = IndexVec::<GeneratorSavedLocal, _>::new();
-    let mut tys = IndexVec::<GeneratorSavedLocal, _>::new();
-    for (saved_local, local) in saved_locals.iter_enumerated() {
-        locals.push(local);
-        tys.push(body.local_decls[local].ty);
-        debug!("generator saved local {:?} => {:?}", saved_local, local);
-    }
-
-    // Leave empty variants for the UNRESUMED, RETURNED, and POISONED states.
+    // Leave variants for the UNRESUMED, RETURNED, and POISONED states.
     // In debuginfo, these will correspond to the beginning (UNRESUMED) or end
     // (RETURNED, POISONED) of the function.
     const RESERVED_VARIANTS: usize = 3;
     let body_span = body.source_scopes[OUTERMOST_SOURCE_SCOPE].span;
-    let mut variant_source_info: IndexVec<VariantIdx, SourceInfo> = [
+    let mut variant_source_info: IndexVec<VariantIdx, SourceInfo> = std::array::IntoIter::new([
         SourceInfo::outermost(body_span.shrink_to_lo()),
         SourceInfo::outermost(body_span.shrink_to_hi()),
         SourceInfo::outermost(body_span.shrink_to_hi()),
-    ]
-    .iter()
-    .copied()
+    ])
     .collect();
 
     // Build the generator variant field list.
     // Create a map from local indices to generator struct indices.
     let mut variant_fields: IndexVec<VariantIdx, IndexVec<Field, GeneratorSavedLocal>> =
         iter::repeat(IndexVec::new()).take(RESERVED_VARIANTS).collect();
+
+    // Gather live local types and their indices.
+    // Store upvars in the UNRESUMED variant.
+    let mut locals = IndexVec::<GeneratorSavedLocal, _>::new();
+    let mut tys = IndexVec::<GeneratorSavedLocal, _>::new();
+    let unresumed_idx = VariantIdx::new(UNRESUMED);
     let mut remap = FxHashMap::default();
+
+    for (saved_local, local) in saved_locals.iter_enumerated() {
+        locals.push(local);
+        tys.push(body.local_decls[local].ty);
+
+        if is_upvar_local(upvars, local) {
+            variant_fields[unresumed_idx].push(saved_local);
+            remap.entry(locals[saved_local]).or_insert((
+                tys[saved_local],
+                unresumed_idx,
+                local.as_usize() - 3,
```

Comment:
Use the length of `variant_fields[unresumed_idx]` to determine the field number.